File: examples/python/vdp_multiple_shooting.py
#
#     This file is part of CasADi.
# 
#     CasADi -- A symbolic framework for dynamic optimization.
#     Copyright (C) 2010 by Joel Andersson, Moritz Diehl, K.U.Leuven. All rights reserved.
# 
#     CasADi is free software; you can redistribute it and/or
#     modify it under the terms of the GNU Lesser General Public
#     License as published by the Free Software Foundation; either
#     version 3 of the License, or (at your option) any later version.
# 
#     CasADi is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#     Lesser General Public License for more details.
# 
#     You should have received a copy of the GNU Lesser General Public
#     License along with CasADi; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
# 
# 
from casadi import *
import numpy as NP
import matplotlib.pyplot as plt

nk = 20      # Control discretization
tf = 10.0    # End time
coll = False # Use collocation integrator

# Declare variables (use scalar graph)
t  = ssym("t")    # time
u  = ssym("u")    # control
x  = ssym("x",3)  # state

# ODE rhs function
ode = vertcat([(1 - x[1]*x[1])*x[0] - x[1] + u, \
       x[0], \
       x[0]*x[0] + x[1]*x[1] + u*u])
f = SXFunction(daeIn(x=x,p=u,t=t),daeOut(ode=ode))

# Create an integrator
if coll:
  f_d = CollocationIntegrator(f)
  f_d.setOption("number_of_finite_elements",5)
  f_d.setOption("interpolation_order",5)
  f_d.setOption("collocation_scheme","legendre")
  f_d.setOption("implicit_solver",KinsolSolver)
  f_d.setOption("implicit_solver_options",\
    {'linear_solver' : CSparse})
  f_d.setOption("expand_f",True)
else:
  f_d = CVodesIntegrator(f)
  f_d.setOption("abstol",1e-8) # tolerance
  f_d.setOption("reltol",1e-8) # tolerance
  f_d.setOption("steps_per_checkpoint",1000)

f_d.setOption("tf",tf/nk) # final time
f_d.init()

# Total number of variables
nv = 1*nk + 3*(nk+1)

# Declare variable vector
V = msym("V", nv)

# Get the expressions for local variables
U = V[0:nk]
X0 = V[nk+0:nv:3]
X1 = V[nk+1:nv:3]
X2 = V[nk+2:nv:3]

# Variable bounds initialized to +/- inf
VMIN = -inf*NP.ones(nv)
VMAX = inf*NP.ones(nv)

# Control bounds
VMIN[0:nk] = -0.75
VMAX[0:nk] = 1.0

# Initial condition
VMIN[nk+0] = VMAX[nk+0] = 0
VMIN[nk+1] = VMAX[nk+1] = 1
VMIN[nk+2] = VMAX[nk+2] = 0

# Terminal constraint
VMIN[nv-3] = VMAX[nv-3] = 0
VMIN[nv-2] = VMAX[nv-2] = 0

# Initial solution guess
VINIT = NP.zeros(nv)

# Constraint function with bounds
g = [];  g_min = []; g_max = []

# Build up a graph of integrator calls
for k in range(nk):
  # Local state vector
  Xk = vertcat((X0[k],X1[k],X2[k]))
  Xk_next = vertcat((X0[k+1],X1[k+1],X2[k+1]))
  
  # Call the integrator
  Xk_end, = integratorOut(f_d.call(integratorIn(x0=Xk,p=U[k])),"xf")
  
  # append continuity constraints
  g.append(Xk_next - Xk_end)
  g_min.append(NP.zeros(Xk.size()))
  g_max.append(NP.zeros(Xk.size()))

# Objective function: L(T)
F = MXFunction([V],[X2[nk]])

# Continuity constraints: 0<= x(T(k+1)) - X(T(k)) <=0
G = MXFunction([V],[vertcat(g)])

# Create NLP solver instance
solver = IpoptSolver(F,G)

#solver.setOption("verbose",True)
solver.init()

# Set bounds and initial guess
solver.setInput(VMIN,  "lbx")
solver.setInput(VMAX,  "ubx")
solver.setInput(VINIT, "x0")
solver.setInput(NP.concatenate(g_min),"lbg")
solver.setInput(NP.concatenate(g_max),"ubg")

# Solve the problem
solver.solve()

# Retrieve the solution
v_opt = solver.output("x")
u_opt = v_opt[0:nk]
x0_opt = v_opt[nk+0::3]
x1_opt = v_opt[nk+1::3]
x2_opt = v_opt[nk+2::3]

# Retrieve the solution
v_opt = NP.array(solver.output("x"))

# Get values at the beginning of each finite element
tgrid_x = NP.linspace(0,10,nk+1)
tgrid_u = NP.linspace(0,10,nk)

# Plot the results
plt.figure(1)
plt.clf()
plt.plot(tgrid_x,x0_opt,'--')
plt.plot(tgrid_x,x1_opt,'-')
plt.plot(tgrid_u,u_opt,'-.')
plt.title("Van der Pol optimization - multiple shooting")
plt.xlabel('time')
plt.legend(['x0 trajectory','x1 trajectory','u trajectory'])
plt.grid()
plt.show()

File: examples/vdp_multiple_shooting.cpp
#include <casadi/fx/fx_tools.hpp>
#include <casadi/mx/mx_tools.hpp>
#include <casadi/sx/sx_tools.hpp>
#include <casadi/matrix/matrix_tools.hpp>
#include <casadi/stl_vector_tools.hpp>

#include <interfaces/ipopt/ipopt_solver.hpp>
#include <interfaces/sundials/cvodes_integrator.hpp>

#include <optimal_control/multiple_shooting.hpp>

using namespace CasADi;
using namespace CasADi::Sundials;
using namespace CasADi::OptimalControl;
using namespace std;

int main(){
  //Final time (fixed)
  double tf = 20.0;

  // Infinity
  double inf = numeric_limits<double>::infinity();

  // Declare variables (use simple, efficient DAG)
  SX t("t"); //time
  SX x("x"), y("y"), u("u"), L("cost");

  //Control bounds
  double u_min[] = {-0.75};
  double u_max[] = {1.0};
  double u_init[] = {0.0};
  
  //State bounds and initial guess
  double x_min[] = {-inf,-inf,-inf};
  double x_max[] = {inf, inf, inf};
  double x_init[] = {0,0,0};

  //State bounds at the initial time
  double x0_min[] = {0,1,0};
  double x0_max[] = {0,1,0};

  //State bounds at the final time
  double xf_min[] = {0,0,-inf};
  double xf_max[] = {0,0, inf};
  
  // All states
  vector<SX> xx(3);  xx[0] = x;  xx[1] = y;  xx[2] = L;

  //ODE right hand side
  vector<SX> f(3);
  f[0] = (1 - y*y)*x - y + u;
  f[1] = x;
  f[2] = x*x + y*y + u*u;
  
  //ODE right hand side function
  vector<SXMatrix> rhs_in(ODE_NUM_IN);
  rhs_in[ODE_T] = t;
  rhs_in[ODE_Y] = xx;
  rhs_in[ODE_P] = u;
  SXFunction rhs(rhs_in,f);

  //Create an integrator (CVodes)
  CVodesIntegrator I(rhs);
  I.setOption("abstol",1e-8); //abs. tolerance
  I.setOption("reltol",1e-8); //rel. tolerance
  I.setOption("steps_per_checkpoint",500);
  I.setOption("stop_at_end",true);
  I.init();
  
  //Numboer of shooting nodes
  int ns = 50;

  // Number of differential states
  int nx = 3;
  
  // Number of controls
  int nu = 1;
  
  // Mayer objective function
  Matrix<SX> xf = symbolic("xf",nx,1);
  SXFunction mterm(xf, xf[nx-1]);

  // Create a multiple shooting discretization
  MultipleShooting ms(I,mterm,ns,nx,nu);
  
  // Copy data
  ms.tf_ = tf;
  
  copy(u_min,u_min+nu,ms.u_min_.begin());
  copy(u_max,u_max+nu,ms.u_max_.begin());
  copy(u_init,u_init+nu,ms.u_init_.begin());
  
  copy(x_min,x_min+nx,ms.x_min_.begin());
  copy(x_max,x_max+nx,ms.x_max_.begin());
  copy(x_init,x_init+nx,ms.x_init_.begin());
  
  copy(x0_min,x0_min+nx,ms.x0_min_.begin());
  copy(x0_max,x0_max+nx,ms.x0_max_.begin());
  
  copy(xf_min,xf_min+nx,ms.xf_min_.begin());
  copy(xf_max,xf_max+nx,ms.xf_max_.begin());

  ms.init();
  
  IpoptSolver solver(ms.F_,ms.G_,FX(),ms.J_);
  solver.setOption("tol",1e-5);
  solver.setOption("hessian_approximation", "limited-memory");
  solver.setOption("max_iter",100);
  solver.setOption("linear_solver","ma57");
//  solver.setOption("derivative_test","first-order");

  //solver.setOption("verbose",true);
  solver.init();

  //Set bounds and initial guess
  solver.setInput(ms.V_min_,  NLP_LBX);
  solver.setInput(ms.V_max_,  NLP_UBX);
  solver.setInput(ms.V_init_,  NLP_X_INIT);
  
  solver.setInput(ms.G_min_,NLP_LBG);
  solver.setInput(ms.G_max_,NLP_UBG);

  //Solve the problem
  solver.solve();

  return 0;
}



